modules/csg_2d/csg_shape_2d.cpp

```diff
@@ -0,0 +1,1046 @@
+				polygon_paths = Clipper2Lib::Difference(polygon_paths, child_polygon_paths, Clipper2Lib::FillRule::NonZero);
+				break;
+			default: {
+				ERR_PRINT("CSGShape2D::Operation failed. Unrecognized operation.");
+				return brush;
+			}
+		}
+	}
+
+	n->poly_paths = polygon_paths;
+
+	Rect2 rect;
+	if (n && !n->outlines.is_empty()) {
+		rect.position = n->outlines[0].vertices[0];
+		for (const CSGBrush2D::Outline &outline : n->outlines) {
+			for (uint32_t i = 0; i < outline.vertices.size(); i++) {
+				rect.expand_to(outline.vertices[i]);
+			}
+		}
+	}
+	node_rect = rect;
+
+	brush = n;
+	dirty = false;
+	return brush;
+}
+
+void CSGShape2D::force_shape_update() {
+	shape_update_forced = true;
+	_update_shape();
+	shape_update_forced = false;
+}
+
+void CSGShape2D::_update_shape() {
+	if (Engine::get_singleton()->is_editor_hint()) {
+		if (brush_mesh.is_null()) {
+			brush_mesh.instantiate();
+		}
+		brush_mesh->clear_surfaces();
+
+		{
+			Array array;
+			array.resize(Mesh::ARRAY_MAX);
+
+			for (const Vector<Vector2> &brush_outline : brush_outlines) {
+				Vector<Vector2> brush_mesh_vertices;
+				brush_mesh_vertices.resize(brush_outline.size() * 2);
+				Vector2 *brush_mesh_vertices_ptrw = brush_mesh_vertices.ptrw();
+
+				for (uint32_t i = 0; i < brush_outline.size(); i++) {
+					const Vector2 &vertex1 = brush_outline[i];
+					const Vector2 &vertex2 = brush_outline[(i + 1) % brush_outline.size()];
+
+					brush_mesh_vertices_ptrw[i * 2] = vertex1;
+					brush_mesh_vertices_ptrw[i * 2 + 1] = vertex2;
+				}
+
+				array[Mesh::ARRAY_VERTEX] = brush_mesh_vertices;
+
+				brush_mesh->add_surface_from_arrays(Mesh::PRIMITIVE_LINES, array, Array(), Dictionary(), RS::ARRAY_FLAG_USE_2D_VERTICES);
+			}
+		}
+	}
+
+	if (!is_root_shape()) {
+		return;
+	}
+
+	mesh_vertices.clear();
+	mesh_triangles.clear();
+	mesh_convex_polygons.clear();
+	mesh_outlines.clear();
+
+	root_mesh.unref();
+	root_edge_line_mesh.unref();
+	root_outline_mesh.unref();
+
+	CSGBrush2D *n = _get_brush();
+	ERR_FAIL_NULL_MSG(n, "Cannot get CSGBrush2D.");
+
+	root_mesh.instantiate();
+	root_edge_line_mesh.instantiate();
+	root_outline_mesh.instantiate();
+
+	Clipper2Lib::PolyTreeD polytree;
+	Clipper2Lib::ClipperD clipper_D;
+	clipper_D.PreserveCollinear(false);
+
+	clipper_D.AddSubject(n->poly_paths);
+	clipper_D.Execute(Clipper2Lib::ClipType::Union, Clipper2Lib::FillRule::NonZero, polytree);
+
+	List<TPPLPoly> tppl_in_polygon, tppl_out_polygon;
+
+	for (size_t i = 0; i < polytree.Count(); i++) {
+		const Clipper2Lib::PolyPathD *polypath_item = polytree[i];
+		_recursive_process_polytree_items(tppl_in_polygon, polypath_item, mesh_outlines);
+	}
+
+	TPPLPartition tpart;
+
+	if (tpart.ConvexPartition_HM(&tppl_in_polygon, &tppl_out_polygon) == 0) {
+		ERR_PRINT("CSGShape2D outines to convex polygons conversion failed. Check input geometry for common shape outline and precision errors like self-overlapping shapes used in intersection operations.");
```

Comment:
To confirm, this isn't a critical situation, it should continue here after failing?

Reply:
It needs to run the code past that point for e.g. redraw so that the visuals actually update with the failed state.